src/compiler/checker.ts

```diff
@@ -13881,6 +13881,10 @@ namespace ts {
                         }
                     }
                 }
+                else if (compilerOptions.noImplicitReturns && !isTypeOfKind(returnType, TypeFlags.Void | TypeFlags.Any)) {

```

Comment:
also looks like we do not handel this correctelly in 10804:
```ts
if (inferredReturnType === voidType || isTypeAny(inferredReturnType)) {
```

should also use `maybeTypeOfKind` as well.